File: go/bsb/events_router.go
package bsb

import (
	"context"
	"fmt"
	"io"
	"time"
)

type eventRoute struct {
	plugin EventsPlugin
	filter any
}
type eventRouter struct{ routes []eventRoute }

func validateEventFilter(filter any) error {
	switch value := filter.(type) {
	case nil:
		return nil
	case []any:
		for _, item := range value {
			if _, ok := item.(string); !ok {
				return fmt.Errorf("event filter lists must contain strings")
			}
		}
	case map[string]any:
		for _, item := range value {
			switch v := item.(type) {
			case bool:
			case []any:
				if err := validateEventFilter(v); err != nil {
					return err
				}
			case map[string]any:
				if _, ok := v["enabled"].(bool); !ok {
					return fmt.Errorf("event filter enabled must be boolean")
				}
				list, ok := v["plugins"].([]any)
				if !ok {
					return fmt.Errorf("event filter plugins must be a list")
				}
				if err := validateEventFilter(list); err != nil {
					return err
				}
			default:
				return fmt.Errorf("invalid event filter operation")
			}
		}
	default:
		return fmt.Errorf("event filter must be a list or object")
	}
	return nil
}
func eventFilterMatches(filter any, operation, plugin string) bool {
	if filter == nil {
		return true
	}
	var list []any
	switch value := filter.(type) {
	case []any:
		for _, item := range value {
			if item == operation {
				return true
			}
		}
		return false
	case map[string]any:
		switch entry := value[operation].(type) {
		case bool:
			return entry
		case []any:
			list = entry
		case map[string]any:
			if entry["enabled"] != true {
				return false
			}
			list, _ = entry["plugins"].([]any)
		}
	}
	for _, item := range list {
		if item == plugin {
			return true
		}
	}
	return false
}
func (r *eventRouter) route(operation, plugin string) (EventsPlugin, error) {
	for _, entry := range r.routes {
		if eventFilterMatches(entry.filter, operation, plugin) {
			return entry.plugin, nil
		}
	}
	return nil, fmt.Errorf("no events backend matches %s for %s", operation, plugin)
}

// The controller owns backend lifecycle; services only use this routing facade.
func (r *eventRouter) Init(context.Context, Observable) error { return nil }
func (r *eventRouter) Run(context.Context, Observable) error  { return nil }
func (r *eventRouter) Dispose() error                         { return nil }
func (r *eventRouter) OnEvent(ctx context.Context, obs Observable, p, e string, h EventListener) error {
	b, err := r.route("onEvent", p)
	if err != nil {
		return err
	}
	return b.OnEvent(ctx, obs, p, e, h)
}
func (r *eventRouter) EmitEvent(ctx context.Context, obs Observable, p, e string, v any) error {
	b, err := r.route("emitEvent", p)
	if err != nil {
		return err
	}
	return b.EmitEvent(ctx, obs, p, e, v)
}
func (r *eventRouter) OnBroadcast(ctx context.Context, obs Observable, p, e string, h BroadcastListener) error {
	b, err := r.route("onBroadcast", p)
	if err != nil {
		return err
	}
	return b.OnBroadcast(ctx, obs, p, e, h)
}
func (r *eventRouter) EmitBroadcast(ctx context.Context, obs Observable, p, e string, v any) error {
	b, err := r.route("emitBroadcast", p)
	if err != nil {
		return err
	}
	return b.EmitBroadcast(ctx, obs, p, e, v)
}
func (r *eventRouter) OnReturnableEvent(ctx context.Context, obs Observable, p, e string, h ReturnableListener) error {
	b, err := r.route("onReturnableEvent", p)
	if err != nil {
		return err
	}
	return b.OnReturnableEvent(ctx, obs, p, e, h)
}
func (r *eventRouter) EmitEventAndReturn(ctx context.Context, obs Observable, p, e string, t time.Duration, v any) (any, error) {
	b, err := r.route("emitEventAndReturn", p)
	if err != nil {
		return nil, err
	}
	return b.EmitEventAndReturn(ctx, obs, p, e, t, v)
}
func (r *eventRouter) ReceiveStream(ctx context.Context, obs Observable, p, e string, h StreamListener, t time.Duration) (string, error) {
	b, err := r.route("receiveStream", p)
	if err != nil {
		return "", err
	}
	return b.ReceiveStream(ctx, obs, p, e, h, t)
}
func (r *eventRouter) SendStream(ctx context.Context, obs Observable, p, e, id string, v io.Reader) error {
	b, err := r.route("sendStream", p)
	if err != nil {
		return err
	}
	return b.SendStream(ctx, obs, p, e, id, v)
}

File: rust/src/events_router.rs
use crate::{
    events::{Bus, Handler, Reader, StreamHandler},
    observable::Observable,
};
use anyhow::{Result, bail, ensure};
use async_trait::async_trait;
use serde_json::Value;
use std::{sync::Arc, time::Duration};

pub(crate) struct Router(pub Vec<(Arc<dyn Bus>, Value)>);
pub(crate) fn validate(filter: &Value) -> Result<()> {
    fn strings(value: &Value) -> bool {
        value
            .as_array()
            .is_some_and(|v| v.iter().all(Value::is_string))
    }
    match filter {
        Value::Null => {}
        Value::Array(_) => ensure!(strings(filter), "event filter lists must contain strings"),
        Value::Object(entries) => {
            for entry in entries.values() {
                ensure!(
                    entry.is_boolean()
                        || strings(entry)
                        || (entry.is_object()
                            && entry["enabled"].is_boolean()
                            && strings(&entry["plugins"])),
                    "invalid event filter operation"
                );
            }
        }
        _ => bail!("event filter must be a list or object"),
    }
    Ok(())
}
fn matches(filter: &Value, operation: &str, plugin: &str) -> bool {
    fn contains(value: &Value, key: &str) -> bool {
        value.as_array().is_some_and(|v| v.iter().any(|v| v == key))
    }
    match filter {
        Value::Null => true,
        Value::Array(_) => contains(filter, operation),
        Value::Object(entries) => match entries.get(operation) {
            Some(Value::Bool(value)) => *value,
            Some(value @ Value::Array(_)) => contains(value, plugin),
            Some(value @ Value::Object(_)) => {
                value["enabled"] == true && contains(&value["plugins"], plugin)
            }
            _ => false,
        },
        _ => false,
    }
}
impl Router {
    fn route(&self, operation: &str, plugin: &str) -> Result<&Arc<dyn Bus>> {
        self.0
            .iter()
            .find(|(_, filter)| matches(filter, operation, plugin))
            .map(|(bus, _)| bus)
            .ok_or_else(|| anyhow::anyhow!("no events backend matches {operation} for {plugin}"))
    }
}
#[async_trait]
impl Bus for Router {
    async fn listen(&self, kind: &str, target: &str, event: &str, handler: Handler) -> Result<()> {
        let operation = match kind {
            "event" => "onEvent",
            "returnable" => "onReturnableEvent",
            "broadcast" => "onBroadcast",
            _ => bail!("unknown event kind"),
        };
        self.route(operation, target)?
            .listen(kind, target, event, handler)
            .await
    }
    async fn emit(
        &self,
        obs: Observable,
        kind: &str,
        target: &str,
        event: &str,
        value: Value,
        timeout: Duration,
    ) -> Result<Value> {
        let operation = match kind {
            "event" => "emitEvent",
            "returnable" => "emitEventAndReturn",
            "broadcast" => "emitBroadcast",
            _ => bail!("unknown event kind"),
        };
        self.route(operation, target)?
            .emit(obs, kind, target, event, value, timeout)
            .await
    }
    async fn receive(
        &self,
        obs: Observable,
        target: &str,
        event: &str,
        handler: StreamHandler,
        timeout: Duration,
    ) -> Result<String> {
        self.route("receiveStream", target)?
            .receive(obs, target, event, handler, timeout)
            .await
    }
    async fn send(
        &self,
        obs: Observable,
        target: &str,
        event: &str,
        id: &str,
        reader: Reader,
    ) -> Result<()> {
        self.route("sendStream", target)?
            .send(obs, target, event, id, reader)
            .await
    }
    // Host owns shutdown and monitors every underlying transport.
    async fn shutdown(&self) -> Result<()> {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{events::LocalBus, json, observable::Backend};
    #[tokio::test]
    async fn filtered_backends_keep_unfiltered_fallback() -> Result<()> {
        let selected: Arc<dyn Bus> = Arc::new(LocalBus::default());
        let fallback: Arc<dyn Bus> = Arc::new(LocalBus::default());
        let filter = json!({"onReturnableEvent":["worker"],"emitEventAndReturn":{"enabled":true,"plugins":["worker"]}});
        validate(&filter)?;
        let router = Router(vec![
            (selected.clone(), filter),
            (fallback.clone(), Value::Null),
        ]);
        let echo: Handler = Arc::new(|_, value| Box::pin(async { Ok(value) }));
        router
            .listen("returnable", "worker", "echo", echo.clone())
            .await?;
        router.listen("returnable", "other", "echo", echo).await?;
        let obs = Observable::new("test", Arc::new(Backend::default()));
        for (bus, target) in [(selected, "worker"), (fallback, "other")] {
            assert_eq!(
                bus.emit(
                    obs.clone(),
                    "returnable",
                    target,
                    "echo",
                    json!(target),
                    Duration::from_secs(1)
                )
                .await?,
                json!(target)
            );
            bus.shutdown().await?;
        }
        assert!(validate(&json!({"emitEvent":42})).is_err());
        assert!(!matches(&json!([]), "onEvent", "worker"));
        Ok(())
    }
}
